pycqed/measurement/waveform_control/segment.py

```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
                 el_new_start = el_list[i + 1][0]
 
                 if (el_prev_end - el_new_start) > tol :
-                    print(el_prev_end, el_new_start)
-                    msg = f'{prev_el} (ends at {el_prev_end*1e6:.4f}us) and ' \
-                    f'{el_list[i + 1][2]} (' \
-                        f'starts at {el_new_start*1e6:.4f}us) overlap ' \
-                          f'on {awg}'
-                    if allow_overlap:
-                        log.warning(msg)
+                    if track_and_ignore:
+                        # add set of two overlapping elements to list
+                        self.overlapping_elements.append({prev_el,
+                                                          el_list[i + 1][2]})
                     else:
-                        raise ValueError(msg)
+                        print(el_prev_end, el_new_start)
+                        msg = f'{prev_el} (ends at {el_prev_end*1e6:.4f}us) ' \
+                              f'and {el_list[i + 1][2]} (' \
+                              f'starts at {el_new_start*1e6:.4f}us) overlap ' \
+                              f'on {awg}'
+                        if allow_overlap:
+                            log.warning(msg)
+                        else:
+                            raise ValueError(msg)
+
+    def resolve_overlap(self):
+        """
+        Routine to resolve overlapping elements. Will be exectued if
+        corresponding pulsar parameter is true. This code
+        first goes through the list of overlapping elements that are
+        pairwise overlapping and clusters them into lists
+        of overlapping elements, where two different lists of overlapping
+        elements have no overlapping elements with
+        one another. At the end the code combines all elements of each
+        list into a new element.
+        """
+
+        self.gen_elements_on_awg()
+        self._test_overlap(track_and_ignore=True)
+        overlapping_elements = self.overlapping_elements
```

Comment:
Why is `self.overlapping_elements` a property? It could just be return value of `_test_overlap` if that method is called with `track_and_ignore=True`.

If there is a reason to have it as a property: should the property be updated to an empty list after resolving the overlaps?

Reply:
I like the above suggestion. Implemented it like you suggested.